spring-cloud-kubernetes-core/src/main/java/org/springframework/cloud/kubernetes/condition/OnKubernetesCondition.java

```diff
@@ -0,0 +1,220 @@
+package org.springframework.cloud.kubernetes.condition;
+
+import static org.springframework.cloud.kubernetes.KubernetesClientProperties.SPRING_CLOUD_KUBERNETES_CLIENT_PROPS;
+
+import io.fabric8.kubernetes.client.Config;
+import io.fabric8.kubernetes.client.ConfigBuilder;
+import io.fabric8.kubernetes.client.DefaultKubernetesClient;
+import io.fabric8.kubernetes.client.KubernetesClient;
+import io.fabric8.kubernetes.client.VersionInfo;
+import java.util.Map;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import org.springframework.boot.autoconfigure.condition.ConditionMessage;
+import org.springframework.boot.autoconfigure.condition.ConditionOutcome;
+import org.springframework.boot.autoconfigure.condition.SpringBootCondition;
+import org.springframework.cloud.kubernetes.condition.ConditionalOnKubernetes.Range;
+import org.springframework.context.annotation.ConditionContext;
+import org.springframework.core.env.PropertyResolver;
+import org.springframework.core.type.AnnotatedTypeMetadata;
+
+class OnKubernetesCondition extends SpringBootCondition {
+
+
+	private static final String ANNOTATION_VERSION_REGEX = "(\\d+)\\.(\\d+)";
+	private static final Pattern GIT_VERSION_PATTERN = Pattern.compile("v(\\d+)\\.(\\d+).*");
+	private static final Pattern ANNOTATION_VERSION_PATTERN =
+		Pattern.compile(ANNOTATION_VERSION_REGEX);
+
+	@Override
+	public ConditionOutcome getMatchOutcome(ConditionContext context,
+		AnnotatedTypeMetadata metadata) {
+
+		ConditionMessage.Builder message = ConditionMessage.forCondition("");
+
+		if (metadata.isAnnotated(ConditionalOnKubernetes.class.getName())) {
+			KubernetesClient kubernetesClient = getKubernetesClient(context.getEnvironment());
+
+			if (!isServerUp(kubernetesClient)) {
+				return ConditionOutcome.noMatch(
+					message.because("Could not communicate with the Kubernetes server at: "
+						+ kubernetesClient.getMasterUrl())
+				);
+			}
+
+			Map<String, Object> attributes = metadata
+				.getAnnotationAttributes(ConditionalOnKubernetes.class.getName());
+
+			ConditionOutcome adaptabilityCheckOutcome = checkAdaptability(
+				(Class<?>) attributes.get("classClientMustAdaptTo"),
+				kubernetesClient
+			);
+
+			if (adaptabilityCheckOutcome.isMatch()) {
+				return checkVersion(
+					(String) attributes.get("version"),
+					(ConditionalOnKubernetes.Range) attributes.get("range"),
+					kubernetesClient
+				);
+			}
+			else {
+				return adaptabilityCheckOutcome;
+			}
+		}
+
+		return ConditionOutcome.noMatch(
+			message.because("Unknown Conditional annotation using " +
+				this.getClass().getSimpleName())
+		);
+	}
+
+	// We need to create a KubernetesClient ourselves
+	// since the KubernetesClient Bean is not available when the Condition is evaluated
```

Comment:
Unfortunately, I don't think this will work very well. The auto-configuration for `KubernetesClient` marks it as `@ConditionalOnMissingBean`. This is a good thing as it allows a user to provide their own `KubernetesClient` configured however they like. This condition effectively breaks that as it assumes that the user will have used the standard configuration properties to configure the client.

Reply:
@wilkinsona That's very good point, thanks for bringing it up!

Unfortunately we have no other way of creating a KubernetesClient in the condition evaluation phase (which is executed well before the user specifies a custom KubernetesClient bean) that I can think of.

I understand that the implementation here is "unusual", in the sense that the output of the `Conditional` depends on some external service. Are you aware of any other Conditional that performs similar functionality?
Is this something that might perhaps make sense to be "standardized" in Spring Boot?